### PR TITLE
version: add the gfapi version check

### PR DESCRIPTION
**Pull request by lxbsz**
Signed-off-by: Xiubo Li <xiubli@redhat.com>

### CODE REVIEW

**Comment by lxbsz** [comment]
@amarts @pkalever 
Please review, thanks.

**Comment by lxbsz** [comment]
```
[root@rhel3 gluster-block]# rpm -qa|grep glusterfs
glusterfs-events-6dev-0.488.gita95157a.el7.x86_64
glusterfs-6dev-0.488.gita95157a.el7.x86_64
glusterfs-server-6dev-0.488.gita95157a.el7.x86_64
glusterfs-cloudsync-plugins-6dev-0.488.gita95157a.el7.x86_64
glusterfs-extra-xlators-6dev-0.488.gita95157a.el7.x86_64
glusterfs-api-devel-6dev-0.488.gita95157a.el7.x86_64
glusterfs-api-6dev-0.488.gita95157a.el7.x86_64
glusterfs-devel-6dev-0.488.gita95157a.el7.x86_64
glusterfs-regression-tests-6dev-0.488.gita95157a.el7.x86_64
glusterfs-gnfs-6dev-0.488.gita95157a.el7.x86_64
glusterfs-libs-6dev-0.488.gita95157a.el7.x86_64
glusterfs-cli-6dev-0.488.gita95157a.el7.x86_64
glusterfs-rdma-6dev-0.488.gita95157a.el7.x86_64
glusterfs-fuse-6dev-0.488.gita95157a.el7.x86_64
glusterfs-geo-replication-6dev-0.488.gita95157a.el7.x86_64
glusterfs-client-xlators-6dev-0.488.gita95157a.el7.x86_64
glusterfs-resource-agents-6dev-0.488.gita95157a.el7.noarch
glusterfs-debuginfo-6dev-0.488.gita95157a.el7.x86_64
[root@rhel3 gluster-block]# grep GFAPI ./config.h
#define GFAPI_VERSION741 1
[root@rhel3 gluster-block]# 

```

File: rpc/glfs-operations.c [hunk 1]
```diff
@@ -212,7 +213,11 @@ glusterBlockCreateEntry(struct glfs *glfs, blockCreateCli *blk, char *gbid,
     ret = -1;
     goto out;
   } else {
+#if GFAPI_VERSION741
+    ret = glfs_ftruncate(tgfd, blk->size, NULL, NULL);
```
Comment (nixpanic) [inline rpc/glfs-operations.c]: I think this change had been introduced with glusterfs-6, not 4.1.

Comment (lxbsz) [reply rpc/glfs-operations.c]: Yeah, right, I make one mistake and just checked this from the tags not the release branches, by using the 'git tag --contains SHA', but this was reverted later :-)

```
commit 248152767b0599986bbb6bb35fc27197f6be6964
Author: Kinglong Mee <kinglongmee@gmail.com>
Date:   Thu Feb 1 10:13:07 2018 +0800

    gfapi: return pre/post attributes from glfs_ftruncate
    
    Updates: #389
    Change-Id: I8faea0828921fb17f05f7321c3cb01747373f21e
    Signed-off-by: Kinglong Mee <mijinlong@open-fs.com>

diff --git a/api/src/glfs.h b/api/src/glfs.h
index 2addacb..e2f2a2c 100644
--- a/api/src/glfs.h
+++ b/api/src/glfs.h
@@ -571,8 +571,9 @@ off_t glfs_lseek (glfs_fd_t *fd, off_t offset, int whence) __THROW
 int glfs_truncate (glfs_t *fs, const char *path, off_t length) __THROW
         GFAPI_PUBLIC(glfs_truncate, 3.7.15);
 
-int glfs_ftruncate (glfs_fd_t *fd, off_t length) __THROW
-        GFAPI_PUBLIC(glfs_ftruncate, 3.4.0);
+int glfs_ftruncate (glfs_fd_t *fd, off_t length, struct stat *prestat,
+                    struct stat *poststat) __THROW
+        GFAPI_PUBLIC(glfs_ftruncate, 4.0.0);
 int glfs_ftruncate_async (glfs_fd_t *fd, off_t length, glfs_io_cbk fn,
                          void *data) __THROW
         GFAPI_PUBLIC(glfs_ftruncate_async, 3.4.0);

commit 09943beb499617212f2985ca8ea9ecd1ed1b470e
[root@rhel3 glusterfs]# git tag --contains 248152767b0599986bbb6bb35fc27197f6be6964
help
v4.1.0
v4.1.0alpha
v4.1.0rc0
v4.1.1
v4.1.2
v4.1.3
v4.1.4
v4.1.5
v4.1.6
v4.1.7
v4.2dev
v5.0
v5.0alpha
v5.0rc0
v5.0rc1
v5.1
v5.2
v5.3
v5.4
v6.0alpha
v6.0rc0
v6.0rc1
v6dev
v7dev
[root@rhel3 glusterfs]# 
```

Will fix it later, thanks.

**Comment by lxbsz** [comment]
@nixpanic Update done, please review and thanks.

File: rpc/Makefile.am [hunk 2]
```diff
@@ -6,9 +6,9 @@ libgbrpc_la_SOURCES = block_svc_routines.c glfs-operations.c
 
 noinst_HEADERS = glfs-operations.h
 
-libgbrpc_la_CFLAGS = $(GFAPI_CFLAGS) $(JSONC_CFLAGS) \
-                       -DDATADIR=\"$(localstatedir)\"  \
-                       -I$(top_srcdir)/utils/ -I$(top_builddir)/rpc/rpcl
+libgbrpc_la_CFLAGS = $(GFAPI_CFLAGS) $(JSONC_CFLAGS)                    \
+					 -DDATADIR=\"$(localstatedir)\" -I$(top_builddir)/  \
```
Comment (amarts) [inline rpc/Makefile.am]: Is this extra tab expected?

Comment (lxbsz) [reply rpc/Makefile.am]: Switched between tcmu-runner and gluster code style, may just forgot to change it back.
Fixed it.
Thanks Amar.

**Comment by pkalever** [comment]
LGTM, will test it and update.

File: rpc/glfs-operations.c [hunk 3]
```diff
@@ -212,7 +213,11 @@ glusterBlockCreateEntry(struct glfs *glfs, blockCreateCli *blk, char *gbid,
     ret = -1;
     goto out;
   } else {
+#if GFAPI_VERSION760
+    ret = glfs_ftruncate(tgfd, blk->size, NULL, NULL);
```
Comment (itisravi) [inline rpc/glfs-operations.c]: Instead of NULL, it is better to add `struct glfs_stat` prestat and poststat buffers.

Comment (lxbsz) [reply rpc/glfs-operations.c]: I checked the source code and it is:

```
2423     if (ret >= 0) {           
2424         if (prestat)          
2425             glfs_iatt_to_statx(glfd->fs, &preiatt, prestat);
2426         if (poststat)         
2427             glfs_iatt_to_statx(glfd->fs, &postiatt, poststat);
2428     }

```
So I just used the NULLs instead currently.

If there is some good reason that we'd better to these buffers, I will update it :-)

Thanks @itisravi .

File: rpc/glfs-operations.c [hunk 4]
```diff
@@ -212,7 +213,11 @@ glusterBlockCreateEntry(struct glfs *glfs, blockCreateCli *blk, char *gbid,
     ret = -1;
     goto out;
   } else {
+#if GFAPI_VERSION760
```
Comment (pkalever) [inline rpc/glfs-operations.c]: @lxbsz 
Just a cosmic suggestion to make it more opt for reader, while the code looks good to me.

```
#ifdef HAVE_LEGACY_FTRUNCATE_GFAPI
    ret = glfs_ftruncate(tgfd, blk->size);
#else
     ret = glfs_ftruncate(tgfd, blk->size, NULL, NULL);
#end
```
Thanks!

Comment (amarts) [reply rpc/glfs-operations.c]: Prasanna, while for this one call it looks good, what if we have to add more `glfs_*` calls later?

Comment (pkalever) [reply rpc/glfs-operations.c]: @amarts AFAIK in 760 ftruncate (and preadv/pwritev which we don't use ATM) signatures got changed ? so if other glfs_* calls change in future we need to redefine GFAPI_VERSION_* again anyway. 

But it was just a cosmic suggestion. I'm okay both way. Feel free to say "lets live with it"

Comment (lxbsz) [reply rpc/glfs-operations.c]: Yeah, this is why I use the 760 instead, if we have new changes again for other APIs.

Comment (pkalever) [reply rpc/glfs-operations.c]: Alright +1 from me too :-)

**Comment by pkalever** [comment]
@lxbsz please update the tags and will be good to get in.

**Comment by lxbsz** [comment]
@pkalever 
Done, thanks.

**Comment by pkalever** [comment]
Thanks @lxbsz

**Pull request merged**